doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+
+So if there is something that is somehow viable outside of these requirements, I probably didn't consider it.
+
+## 4. Solution
+
+Imports, instances and exports are resolved by the name resolver. This is where any missing or conflicting name errors should be flagged. Flattening assumes all names are present and not conflicting.
+
+### Changes to name collection
+
+Although we don't need or want to change anything in the name resolution phase for flattening, we do collect some additional metadata in this phase to be used later in flattening
+
+#### `LookupTable` with full quint definitions
+
+Our current lookup table only stores a projection of the quint definition: id (called reference), name and type annotation (if present). There is no strong reason for that, only a premature optimization and the fact that the lookup table values might also be lambda parameters, which are not quint definitions.
+
+In flattening, we need to manipulate and reorganize definitions, and is much better for performance and readability if the definitions are in the lookup table, as oppose to having to scan the modules for a definition.
+
+Therefore, we change the lookup table to have either a `QuintDef` or a `QuintLambdaParam` as its value, as long as two additional fields (`namespaces` and `importedFrom`) described below.
+
+#### `namespaces`
+
+When collecting names from import/export/instance statements, we should be able to keep track of the namespaces to add to the definition in order to refer back to it uniquely in a flattened module. The following method describes which namespaces to add depending on `def`, returning a list of namespaces to be accumulated starting from the namespace closer to the actual name (i.e. namespaces `['a', 'b', 'c']` applied to name `foo` will result in `c::b::a::foo`).
+
+```typescript
+private namespaces(def: QuintImport | QuintInstance | QuintExport): string[] {
+  if (def.kind === 'instance') {
+    return compact([def.qualifiedName ?? def.protoName, this.currentModuleName])
+  }
+
+  return def.defName ? [] : [def.qualifiedName ?? def.protoName]
+}
+```
+
+##### Namespaces for imports
+
+Imports are the simplest scenario, since they only copy definitions, possibly adding a namespace to them:
+1. `import A.*`, `import A.foo` don't add namespaces
+2. `import A` adds `A::` namespace
+3. `import A as MyA` adds `MyA` namespace
+
+##### Namespaces for instances
+
+For instances, we need to ensure uniqueness, since the same names in the instanced module can assume different values if the constants are different. Therefore, we add `this.currentModuleName` to the namespace, differentiating possible instances with the same name from different modules. For the following cases, suppose the statements are inside the `myModule` module.
+1. `import A(N=1).*`  adds a `myModule::A` namespace
+2. `import A(N=1) as A1` adds a `myModule::A1` namespace
```

Comment:
Not clear to me how the examples here play out. If I `import A(N=1).*`, are the names from `A` available inside `myModule` in an unqualified version?
* If yes, is this because there is always an implicit `myModule::` qualification?
* If no, I'm afraid that this asymmetry compared to "regular" imports may be confusing to users.

Reply:
I think you are mixing things up. Name "availability" is part of name resolution, and yes, `A` will be available inside `myModule` unqualified. This is what the user sees. Then, after the user got all their names correctly, the modules will go through flattening, which might add namespaces to both the definition and its references. The user should never see this additional namespaces, and they are only meant to disambiguate names in the flattened module, used by the compiler and apalache.